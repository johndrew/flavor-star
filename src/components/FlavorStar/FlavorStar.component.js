import React, { Component } from 'react';
import FlavorCircle from '../FlavorCircle/FlavorCircle.component';

export default class FlavorStar extends Component {

    render() {

        const width = 640;
        const height = width;
        const radius = width * 0.109375;
        const length = width * 0.625;
        const {
            bottom_left,
            bottom_right,
            top,
            top_left,
            top_right,
        } = this.getCoordinates({
            x: width * 0.109375,
            y: width * 0.3125,
            length,
        });

        return (
            <div>
                <svg
                    height={height}
                    width={width}>
                    <FlavorCircle type="salty_umami" x={top.x} y={top.y} radius={radius} />
                    <FlavorCircle type="sweet" x={top_right.x} y={top_right.y} radius={radius} />
                    <FlavorCircle type="sour" x={bottom_right.x} y={bottom_right.y} radius={radius} /> */}
                    <FlavorCircle type="bitter" x={bottom_left.x} y={bottom_right.y} radius={radius} />
                    <FlavorCircle type="spice" x={top_left.x} y={top_left.y} radius={radius} />
                </svg>
            </div>
        );
    }

    /**
     * Determines the coordinates of the star based on specific size parameters
     * Logic pulled from https://rosettacode.org/wiki/Pentagram#Java
     * @typedef {object} Coordinates
     * @prop {number} x
     * @prop {number} y
     * @typedef {object} StarCoordinates
     * @prop {Coordinates} bottom_left The bottom left point of the star
     * @prop {Coordinates} bottom_right The bottom right point of the star
     * @prop {Coordinates} top The top point of the star
     * @prop {Coordinates} top_left The top left point of the star
     * @prop {Coordinates} top_right The top right point of the star
     * @param {number} [angle] An angle in radians
     * @param {number} length The length of the sides
     * @param {number} radius The radius of the star, from center to top point, if a circle were drawn around it
     * @returns {StarCoordinates} Cartesian coordinates to draw the star
     */
    getCoordinates({ x = 0, y = 0, length, angle = 0 }) {
        // TODO: remove x,y defaults and update tests

        let calcAngle = angle;

        // Points start with the top left
        const points = [{ x, y}];
        for (var i = 0; i < 5; i += 1) {

            // This calculation will move in this order: top left, top right, bottom right, top, bottom left
            const point = {
                x: x + Math.cos(calcAngle) * length,
                y: y + Math.sin(-calcAngle) * length,
            };
            points.push(point);
            ({ x, y } = point);
            calcAngle -= 144 * (Math.PI / 180); // 144 degrees in radians
        }
        return {
            bottom_left: points[4],
            bottom_right: points[2],
            top: points[3],
            top_left: points[0],
            top_right: points[1],
        };
    }
}