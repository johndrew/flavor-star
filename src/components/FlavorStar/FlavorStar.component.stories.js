import React from 'react';
import { storiesOf } from '@storybook/react';
import FlavorStar from './FlavorStar.component';

storiesOf('FlavorStar', module)
    .add('without props', flavorStarWithoutProps)
    .add('small scale', flavorStarSmallScale);

function flavorStarWithoutProps() {
    
    return <FlavorStar />;
}

function flavorStarSmallScale() {
    
    const props = {
        width: 320,
    };
    return <FlavorStar {...props}/>;
}