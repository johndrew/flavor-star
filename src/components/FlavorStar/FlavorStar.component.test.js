import React from 'react';
import { shallow } from 'enzyme';
import FlavorStar from './FlavorStar.component';

describe('<FlavorStar />', () => {

    it('renders correctly', () => {

        // TODO: learn why mount caused Jest te4st runner to stall
        expect(shallow(<FlavorStar />)).toMatchSnapshot();
    });

    describe('method getCoordinates', () => {

        let instance;
        beforeEach(() => {
            
            instance = shallow(<FlavorStar />).instance();
        });
        
        describe('Positive Tests', () => {
            
            it('it should return correct coordinates', () => {
                
                const x = 100;
                const y = 0;
                const length = 100;
                const actual = instance.getCoordinates({ x, y, length });
                expect(actual.top_left).toEqual({ x: 100, y: 0 });
                expect(actual.top_right).toEqual({ x: 200, y: 0 });
                expect(actual.bottom_right).toEqual({ x: 119.09830056250527, y: 58.77852522924732 });
                expect(actual.top).toEqual({ x: 150, y: -36.32712640026804 });
                expect(actual.bottom_left).toEqual({ x: 180.90169943749478, y: 58.77852522924731 });
            });
        });
    });
});