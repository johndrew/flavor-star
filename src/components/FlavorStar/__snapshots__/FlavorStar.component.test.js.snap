// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlavorStar /> renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FlavorStar />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <svg
        height={640}
        width={640}
      >
        <FlavorCircle
          radius={70}
          type="salty_umami"
          x={270}
          y={54.69149439892783}
        />
        <FlavorCircle
          radius={70}
          type="sweet"
          x={470}
          y={200}
        />
        <FlavorCircle
          radius={70}
          type="sour"
          x={146.3932022500211}
          y={435.1141009169893}
        />
         */}
        <FlavorCircle
          radius={70}
          type="bitter"
          x={393.6067977499791}
          y={435.1141009169893}
        />
        <FlavorCircle
          radius={70}
          type="spice"
          x={70}
          y={200}
        />
      </svg>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <FlavorCircle
            radius={70}
            type="salty_umami"
            x={270}
            y={54.69149439892783}
          />,
          <FlavorCircle
            radius={70}
            type="sweet"
            x={470}
            y={200}
          />,
          <FlavorCircle
            radius={70}
            type="sour"
            x={146.3932022500211}
            y={435.1141009169893}
          />,
          " */}",
          <FlavorCircle
            radius={70}
            type="bitter"
            x={393.6067977499791}
            y={435.1141009169893}
          />,
          <FlavorCircle
            radius={70}
            type="spice"
            x={70}
            y={200}
          />,
        ],
        "height": 640,
        "width": 640,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "radius": 70,
            "type": "salty_umami",
            "x": 270,
            "y": 54.69149439892783,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "radius": 70,
            "type": "sweet",
            "x": 470,
            "y": 200,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "radius": 70,
            "type": "sour",
            "x": 146.3932022500211,
            "y": 435.1141009169893,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        " */}",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "radius": 70,
            "type": "bitter",
            "x": 393.6067977499791,
            "y": 435.1141009169893,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "radius": 70,
            "type": "spice",
            "x": 70,
            "y": 200,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "svg",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <svg
          height={640}
          width={640}
        >
          <FlavorCircle
            radius={70}
            type="salty_umami"
            x={270}
            y={54.69149439892783}
          />
          <FlavorCircle
            radius={70}
            type="sweet"
            x={470}
            y={200}
          />
          <FlavorCircle
            radius={70}
            type="sour"
            x={146.3932022500211}
            y={435.1141009169893}
          />
           */}
          <FlavorCircle
            radius={70}
            type="bitter"
            x={393.6067977499791}
            y={435.1141009169893}
          />
          <FlavorCircle
            radius={70}
            type="spice"
            x={70}
            y={200}
          />
        </svg>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <FlavorCircle
              radius={70}
              type="salty_umami"
              x={270}
              y={54.69149439892783}
            />,
            <FlavorCircle
              radius={70}
              type="sweet"
              x={470}
              y={200}
            />,
            <FlavorCircle
              radius={70}
              type="sour"
              x={146.3932022500211}
              y={435.1141009169893}
            />,
            " */}",
            <FlavorCircle
              radius={70}
              type="bitter"
              x={393.6067977499791}
              y={435.1141009169893}
            />,
            <FlavorCircle
              radius={70}
              type="spice"
              x={70}
              y={200}
            />,
          ],
          "height": 640,
          "width": 640,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "radius": 70,
              "type": "salty_umami",
              "x": 270,
              "y": 54.69149439892783,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "radius": 70,
              "type": "sweet",
              "x": 470,
              "y": 200,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "radius": 70,
              "type": "sour",
              "x": 146.3932022500211,
              "y": 435.1141009169893,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          " */}",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "radius": 70,
              "type": "bitter",
              "x": 393.6067977499791,
              "y": 435.1141009169893,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "radius": 70,
              "type": "spice",
              "x": 70,
              "y": 200,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "svg",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
